perf_benchmark/README.md

```diff
@@ -42,9 +42,145 @@ Options meaning:
 
 - Elapsed time and CPU time
 - Byte latency and throughput
-- Message latency and throughput
+- Message latency and throughput (message latency should equal to CPU time).
 - Request latency and throughput
 
 We also capture p25, p50, p75, p90, p99, and p999 for each test,
 but `--benchmark_repetitions=1000` is recommended for the results to be
-meaningful.
\ No newline at end of file
+meaningful.
+
+## Run in docker
+
+We use [rules_docker](https://github.com/bazelbuild/rules_docker) to package the
+benchmark binary into a docker image. To build it
+
+```bash
+# change `bazel build` to `bazel run` to start the container directly
+bazel build //perf_benchmark:benchmark_main_image --compilation_mode=opt
+```
+
+There is also a `benchmark_main_image_push` rule to push the image to a docker
+registry.
+
+```bash
+bazel run //perf_benchmark:benchmark_main_image_push \
+ --define=PUSH_REGISTRY=gcr.io \
+ --define=PUSH_PROJECT=project-id \
+ --define=PUSH_TAG=latest
+```
+
+## Benchmark Results
+
+### Environment Setup
+
+We ran the benchmark on the `n1-highmem-32` machines offered from Google Cloud
+Kubernetes Engine (GKE). The container is running in Debian 12.
+
+The request memory and cpu are `512Mi` and `500m` respectively, and the limit
+memory and cpu are `2Gi` and `2` respectively. The transcoder and benchmark
+binaries run in a single-thread, so vCPU with 2 cores is sufficient.
+
+The benchmark was started using the following arguments
+
+```
+--benchmark_repetitions=1100 \
+--benchmark_counters_tabular=true \
+--benchmark_min_warmup_time=3 \
+--benchmark_format=csv
+```
+
+Below, we present the visualization using the median values among the 1100
+repetitions.
+
+### Number of Nested Layers
+
+There are two ways to represent nested JSON structure - using a recursively
+defined protobuf message (`NestedPayload` in `benchmark.proto`) or
+using `google/protobuf/struct.proto`. We benchmarked the effects of having
+nested layers during the transcoding.
+
+- The performance of `google/protobuf/struct.proto` is worse than a recursively
+  defined protobuf message in both JSON -> gRPC and gRPC -> JSON cases.
+- Transcoding for gRPC -> JSON has a much better performance.
+- Transcoding streamed messages doesn't add extra overhead.
+- The per-byte latency of nested structure does not conform to a trend.
+
+![Number of Nested Layers Visualization](image/nested_layers.jpg "Number of Nested Layers")
```

Comment:
Just want to confirm i understand this correctly:

In the [Streaming] case, the y axis is latency per message, right? So this shows that streaming RPCs does not add any overhead (no matter how many messages are streamed, the per-message latency is the same, latency does not depend on number of messages).

Reply:
Yes, you're correct! I have separated out `request latency` and `message latency` during the benchmarks. For the non-streaming case, they are the same. However, for the streaming case, `request latency` = `message latency` * `number of streamed messages`. I didn't present `request latency` in the results because they don't offer more insights than the `message latency`.

I will be more specific in my wordings - there is no extra overhead per streamed message. The number of streamed messages does not affect the per-message latency. Thanks for the insights!